common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -381,32 +384,54 @@ void initialize(String hostname, OpenSSLKey channelIdPrivateKey) throws IOExcept
     // TODO(nathanmittler): Remove once after we switch to the engine socket.
     void doHandshake(FileDescriptor fd, int timeoutMillis)
             throws CertificateException, IOException {
-        if (isClosed() || fd == null || !fd.valid()) {
-            throw new SocketException("Socket is closed");
+        lock.readLock().lock();
+        try {
+            if (isClosed() || fd == null || !fd.valid()) {
+                throw new SocketException("Socket is closed");
+            }
+            NativeCrypto.SSL_do_handshake(ssl, this, fd, handshakeCallbacks, timeoutMillis);
+        } finally {
+            lock.readLock().unlock();
         }
-        NativeCrypto.SSL_do_handshake(ssl, this, fd, handshakeCallbacks, timeoutMillis);
     }
 
     int doHandshake() throws IOException {
-        return NativeCrypto.ENGINE_SSL_do_handshake(ssl, this, handshakeCallbacks);
+        lock.readLock().lock();
+        try {
+            return NativeCrypto.ENGINE_SSL_do_handshake(ssl, this, handshakeCallbacks);
```

Comment:
Why do some methods have the isClosed check whereas others do not ? (other examples, readDirectByteBuffer, writeDirectByteBuffer and forceRead)

Reply:
The pattern is that the methods used by SSLSocket have the check and the ones used by SSLEngine don't, but they should probably all check for being closed.  (Note that it's a case of which exception is thrown rather than correctness, as the native code checks for ssl == 0 and throws NullPointerException in that case.  It'd be better to throw SSLException if it's closed.)  Another change I'd like to push to a future change.